autoprecompiles/src/optimizer.rs

```diff
@@ -112,6 +118,119 @@ fn solver_based_optimization<T: FieldElement>(
+            .map(|(_, constraint)| constraint)
+            .collect(),
+    }
+}
+
+fn can_always_solve<T: FieldElement>(
+    constraint_system: &ConstraintSystem<T, Variable>,
+    constraint_system_item: &ConstraintSystemItem,
+    free_variable: &Variable,
+    bus_interaction_handler: impl BusInteractionHandler<T> + Clone + 'static,
+) -> bool {
+    match constraint_system_item {
+        ConstraintSystemItem::AlgebraicConstraint(i) => {
+            // If an algebraic constraint is linear in an otherwise unconstrained variable,
+            // the prover could always satisfy it by solving for the free variable.
+            is_linear_in_variable(&constraint_system.algebraic_constraints[*i], free_variable)
+        }
+        ConstraintSystemItem::BusInteraction(i) => {
+            let bus_interaction = &constraint_system.bus_interactions[*i];
+
+            let Ok(arg) = bus_interaction
+                .payload
+                .iter()
+                .filter(|payload| payload.try_to_number().is_none())
+                .exactly_one()
+            else {
+                // If there is more non-concrete arguments, there could be an inter-dependence
+                // between them, which would be complicated, so let's skip that for now.
+                // If there are zero non-concrete arguments, the interaction will be removed
+                // by the `remove_trivial_bus_interactions` step.
+                return false;
+            };
+
+            // If the non-concrete argument is linear in the free variable, we know that
+            // we could solve for any valid argument of the bus interaction.
+            if !is_linear_in_variable(arg, free_variable) {
+                // It could still be always solvable, but too complicated...
+                return false;
+            }
+
+            let range_constraints = BusInteraction::from_iter(bus_interaction.iter().map(|expr| {
+                if let Some(value) = expr.try_to_number() {
+                    RangeConstraint::from_value(value)
+                } else {
+                    RangeConstraint::default()
+                }
+            }));
+
+            // If this returns an error, then this bus interaction is actually unsatisfiable.
+            bus_interaction_handler
+                .handle_bus_interaction_checked(range_constraints)
+                .is_ok()
+        }
+    }
+}
+
+fn is_linear_in_variable<T: FieldElement>(
+    constraint: &QuadraticSymbolicExpression<T, Variable>,
+    free_variable: &Variable,
+) -> bool {
+    let (quadratic, linear, _constant) = constraint.components();
+
+    let quadratic_variables = quadratic
+        .iter()
+        .flat_map(|(left, right)| {
+            left.referenced_variables()
+                .chain(right.referenced_variables())
+        })
+        .collect::<HashSet<_>>();
+    if quadratic_variables.contains(free_variable) {
+        // TODO: It could still be linear in the free variable!
```

Comment:
In which way? Because things cancel out or because `right` is a constant?

Reply:
`X * Y * Z` is still linear in `X`, and it would be sound to remove it. What's not sound is, for example `X * X - Y` (constraints `Y` to be a quadratic residue).

Reply:
BTW, one common non-linear constraint would be `X * (X - 1)`. Maybe we should add another rule that we remove any algebraic constraint that *only* contains an otherwise free variable. Although in theory, it could be that the constraint is just not satisfiable (e.g. `X * X - <non quadratic residue>`), so we'd have to assume that there is no contradiction in the input constraint system...

But anyway, this PR seems to be enough to undo the effect of #2717, so I don't think I'll try to be smart in this PR.